update rpc tensorpipe logic for sparse tensors

gcramer23 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62960

A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates rpc/tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

This pr also updates ivalue.cpp and ivalue.h to support more paths for Sparse COO tensors.

I tested these changes by adding sparse tests to rpc_test.py and dist_autograd_test.py.

Differential Revision: [D30717285](https://our.internmc.facebook.com/intern/diff/D30717285)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

gcramer23 committed:
update rpc tensorpipe logic for sparse tensors

[ghstack-poisoned]

mrshenli reviewed (commented):
Would I be correct if I assume this change will no longer be necessary if the SparseTensor logic is added to RPC pickler instead?

gcramer23 commented:
> Would I be correct if I assume this change will no longer be necessary if the SparseTensor logic is added to RPC pickler instead?

Yes because this changes are based on making a sparse tensor a picklable object using JIT pickler.

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

I tested these changes by add sparse checks uncoalesced and coalesced to TensorPipeAgentCudaRpcTest in rpc_test.py 

[ghstack-poisoned]

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

I tested these changes by add sparse checks uncoalesced and coalesced to TensorPipeAgentCudaRpcTest in rpc_test.py 

[ghstack-poisoned]

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

I tested these changes by add sparse checks uncoalesced and coalesced to TensorPipeAgentCudaRpcTest in rpc_test.py 

[ghstack-poisoned]

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

I tested these changes by add sparse checks uncoalesced and coalesced to TensorPipeAgentCudaRpcTest in rpc_test.py 

[ghstack-poisoned]

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates rpc/tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

This pr also updates ivalue.cpp and ivalue.h to support more paths for Sparse COO tensors.

I tested these changes by adding sparse tests to rpc_test.py and dist_autograd_test.py.

[ghstack-poisoned]

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates rpc/tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

This pr also updates ivalue.cpp and ivalue.h to support more paths for Sparse COO tensors.

I tested these changes by adding sparse tests to rpc_test.py and dist_autograd_test.py.

[ghstack-poisoned]

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates rpc/tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

This pr also updates ivalue.cpp and ivalue.h to support more paths for Sparse COO tensors.

I tested these changes by adding sparse tests to rpc_test.py and dist_autograd_test.py.

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -311,8 +311,9 @@ c10::intrusive_ptr<Message> tensorpipeDeserialize(
 
-  for (const auto i : c10::irange(tpDescriptor.tensors.size())) {
-    auto& tensor = tpDescriptor.tensors[i];
+  size_t tpDescriptorPos = 0;
```
Pos->Index 

To stay consistent with the rest of the file, e.g., 

https://github.com/pytorch/pytorch/blob/22bae0d12c2c1f645bb8d0bb95ad3b33b500e343/torch/csrc/distributed/rpc/tensorpipe_utils.cpp#L296

gcramer23 replied:
Okay.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_utils.cpp:
```diff
@@ -326,6 +327,11 @@ c10::intrusive_ptr<Message> tensorpipeDeserialize(
           ", but got it on ",
           tensors[i].device());
     }
+    if (tensors[i].is_sparse()) {
+      tpDescriptorPos += 2;
+    } else {
+      tpDescriptorPos += 1;
+    }
```
will it be safer if we test both indices and values tensor descriptors, instead of just testing values?

gcramer23 replied:
Testing both is unnecessary because if the first fails, then the second falls. The additional tensor descriptor for sparse tensors does not change the order that the descriptors are sent.

Tensors -> (dense1,dense2,sparse3)
Descriptors -> (dense1,dense2,sparse3,sparse3)

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -78,14 +78,20 @@ def create_tensor():
         tensor = tensor.coalesce()
     return tensor
 
+def build_sparse_one_gradient(dtype=torch.float32):
```
looks like this is only used once, maybe does not worth to make it a function?

gcramer23 replied:
It is used two times. I would have to add the def to the test, so I am going to keep it.

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2525,25 +2531,44 @@ def test_thread_local_context_id(self):
-        t2 = torch.rand(3, 3, requires_grad=True, device="cuda:0")
-        (t1 + t2).sum().backward()
+
+    def _gpu_simple(self, t1, t2, sparse):
```
`sparse` is redundant as `t1.is_sparse()` can also tell that?

gcramer23 replied:
Yes, I will update the file.

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2712,17 +2796,26 @@ def test_dist_autograd_sync_streams(self):
         for _ in range(10):
-            input = torch.rand([1000, 10000], device=self.rank, requires_grad=True)
+            if sparse:
+                input = build_sparse_tensor(requires_grad=True, device=self.rank)
```
can be done in followup PR: shall we try larger scale scale tensors with randomly generated indices and values? Looks like, the dense part of this test tries 1000 X 10000, but the sparse part only tries 3 elements?

gcramer23 replied:
That makes sense. It might make sense in the future to update tests that use random values instead to use a set of combinations. For example, the set can be generated using the boundary testing technique.

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2712,17 +2796,26 @@ def test_dist_autograd_sync_streams(self):
             with dist_autograd.context() as context_id:
                 result = local_compute(input)
-                loss = result.sum() * r
+                if sparse:
```
This if-sparse-else has occur many times in this test file, shall we either extract that to a function or simplify it as 

```python
def my_sum(t):
  return torch.sparse.sum(result) if result.is_sparse else result.sum()

# or

torch.sparse.sum(result) if result.is_sparse else result.sum()
```

gcramer23 replied:
I'll extract it to a function.

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2756,8 +2860,8 @@ def test_gradients_synchronizations(self):
-            x = torch.randn(5000, 2000).to(0)
             # local iteration
             local_model = nn.Sequential(*local_layers)
             local_model(x).sum().backward()
```
curious, IIUC, x can be a sparse tensor here and `local_model` contains a sequence of dense layers. This means we can directly pass sparse tensors to dense layers?

gcramer23 replied:
Yes.

```python
class MyModel(torch.nn.Module):

    def __init__(self):
        super().__init__()
        self.fc = torch.nn.Linear(4, 4)

    def forward(self, x):
        return self.fc(x)

if __name__ == "__main__":
    x = torch.rand(4,4).to_sparse()
    m = MyModel()
    out = m(x)
    print(out)
 ```

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2747,7 +2847,11 @@ def test_gradients_synchronizations(self):
             # this is master
-            layers = [nn.Linear(2000, 2000) for _ in range(self.world_size - 1)]
+            if sparse:
+                linear_size = 3
```
shall we keep the linear size and change the input sparse tensor size?

gcramer23 replied:
I changed the input to the function for the sparse test. 

```
    @skip_if_lt_x_gpu(4)
    def test_gradients_synchronizations_sparse(self):
        self._gradients_synchronizations(
            torch.randn(5000, 2000).to_sparse()
        )
 ```

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -2779,3 +2883,17 @@ def test_gradients_synchronizations(self):
+    def test_gradients_synchronizations_sparse(self):
+        self._gradients_synchronizations(
+            build_sparse_tensor(requires_grad=False),
+            True
```
`sparse=True` is redundant

mrshenli reviewed (approved):
LGTM! Added some minor comments. Thanks!

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates rpc/tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

This pr also updates ivalue.cpp and ivalue.h to support more paths for Sparse COO tensors.

I tested these changes by adding sparse tests to rpc_test.py and dist_autograd_test.py.

[ghstack-poisoned]

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates rpc/tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

This pr also updates ivalue.cpp and ivalue.h to support more paths for Sparse COO tensors.

I tested these changes by adding sparse tests to rpc_test.py and dist_autograd_test.py.

[ghstack-poisoned]

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates rpc/tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

This pr also updates ivalue.cpp and ivalue.h to support more paths for Sparse COO tensors.

I tested these changes by adding sparse tests to rpc_test.py and dist_autograd_test.py.

[ghstack-poisoned]

gcramer23 commented:
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30717285).

gcramer23 committed:
Update on "update rpc tensorpipe logic for sparse tensors"


A bug was filed a few years ago for sending sparse tensor over rpc #30807.

This pr updates rpc/tensorpipe logic for CUDA sparse tensors. During the serialization process, the pickler.cpp implementation breaks down the sparse tensor into two tensors and metadata. torch/csrc/distributed/rpc/tensorpipe_agent.cpp needs to be updated because it does not have logic sparse tensors. It pushes a single device for a sparse tensor. This is wrong because after the sparse tensor has been serialized, there will be two tensors. The second tensor will not have a device. This will cause the second tensor to have the wrong target device. tensorpipe_utils.cpp needs to be updated because deserialization happens after the data is received on the target pipe. This takes the two tensors and metadata sent and rebuilds the sparse tensor. There will be two tpDescriptors but only one tensor after deserialization. The logic is updated to verify the sparse tensor is on the correct device using the first tpDescriptor.

This pr also updates ivalue.cpp and ivalue.h to support more paths for Sparse COO tensors.

I tested these changes by adding sparse tests to rpc_test.py and dist_autograd_test.py.

Differential Revision: [D30717285](https://our.internmc.facebook.com/intern/diff/D30717285)

[ghstack-poisoned]

gcramer23 commented:
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30717285).

zhaojuanmao commented:
caused flaky tests, reverting it, see https://github.com/pytorch/pytorch/issues/64518
